php-frontend/src/main/java/org/sonar/php/tree/symbols/SymbolVisitor.java

```diff
@@ -551,29 +549,50 @@ private void usageForNamespaceName(NamespaceNameTree namespaceName, Symbol.Kind
 
   @Override
   public void visitMemberAccess(MemberAccessTree tree) {
-    memberStatementDepth++;
-    if (memberStatementDepth >= MEMBER_ACCESS_MAXIMUM_DEPTH) {
-      if (LOG.isDebugEnabled()) {
-        LOG.debug("visitMemberAccess at \"{}...\" has reached the maximum depth of {}, switching to the next statement",
-          tree.toString().substring(0, MEMBER_ACCESS_LOG_TRUNCATION_LENGTH), MEMBER_ACCESS_MAXIMUM_DEPTH);
-      }
-      return;
+    boolean isCallChain = tree.object().is(Kind.FUNCTION_CALL);
+
+    if (isCallChain) {
+      // special handling of long call chains which happens in some DSL-like scenarios, including but not limited to SQL queries.
+      // Because we unwind the chain from the end, we don't know its exact length, and this branch will execute even for short chains.
+      Stream.iterate(tree, Objects::nonNull, (Tree next) -> {
+        if (next instanceof MemberAccessTree memberAccessTree) {
+          return memberAccessTree.object();
+        } else if (next instanceof FunctionCallTree functionCallTree) {
+          return functionCallTree.callee();
+        } else {
+          return null;
+        }
+      })
+        .forEach((Tree next) -> {
+          if (next instanceof MemberAccessTree memberAccessTree) {
+            visitMemberAccess(memberAccessTree, false);
+          } else if (next instanceof FunctionCallTree functionCallTree) {
+            visitFunctionCall(functionCallTree, false);
+          } else if (!next.is(Kind.NAMESPACE_NAME)) {
+            // namespace name is handled in a corresponding `visit*` method
+            next.accept(this);
+          }
```

Comment:
I'm a bit puzzled by this part.
According to the iterate logic above, we stop and don't process items that are neither `MemberAccessTree` nor `FunctionCallTree`, except for the first element, which we process in any case.
So this `else if(...)` body can only be reached if the first element, `Tree next`, is not a `MemberAccessTree`/`FunctionCallTree`.
Is it expected?
In any case, this use of `Stream.iterate(...)` tends to obfuscate the logic, at least to me :sweat_smile:

Reply:
This will actually happen for the _last_ element, i.e. either the last `.object()` or the last `.callee()`, that is neither Function Call nor Member Access. So basically, this branch calls the rest of processing after the chain is done